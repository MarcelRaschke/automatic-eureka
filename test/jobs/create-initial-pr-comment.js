const nock = require('nock')
const dbs = require('../../lib/dbs')
const upsert = require('../../lib/upsert')
const removeIfExists = require('../helpers/remove-if-exists')

const timeToWaitAfterTests = 500
const waitFor = (milliseconds) => {
  return new Promise((resolve) => {
    setTimeout(resolve, milliseconds)
  })
}

nock.disableNetConnect()
nock.enableNetConnect('localhost')

const createInitialPrComment = require('../../jobs/create-initial-pr-comment')
describe('create-initial-pr-comment', async () => {
  beforeEach(async () => {
    const { installations, repositories } = await dbs()

    await installations.put({
      _id: '123',
      installation: 37
    })
    await repositories.put({
      _id: '42',
      id: '42',
      accountId: '123',
      enabled: false,
      fullName: 'finnp/test',
      owner: {
        id: 123
      }
    })
    await repositories.put({
      _id: '42:branch:deadbeef',
      type: 'branch',
      initial: true,
      sha: 'deadbeef',
      base: 'master',
      head: 'greenkeeper/initial',
      processed: false,
      depsUpdated: true,
      travisModified: true,
      badgeAdded: true,
      badgeUrl: 'url'
    })
    await repositories.put({
      _id: '42:pr:1234',
      repositoryId: '42',
      accountId: '123',
      type: 'pr',
      initial: true,
      number: 1234,
      head: 'greenkeeper/initial',
      state: 'open'
    })
  })

  afterEach(async () => {
    nock.cleanAll()
    const { installations, repositories } = await dbs()
    await Promise.all([
      removeIfExists(installations, '123'),
      removeIfExists(repositories, '42', '42:branch:deadbeef', '42:pr:1234')
    ])
  })

  test('create comment for initial pr created by user', async () => {
    expect.assertions(2)

    nock('https://api.github.com')
      .post('/installations/123/access_tokens')
      .optionally()
      .reply(200, {
        token: 'secret'
      })
      .get('/rate_limit')
      .optionally()
      .reply(200, {})
      .get('/repos/finnp/test')
      .reply(200, {
        name: 'test',
        html_url: 'https://github.com/finnp/test',
        newBranch: 'greenkeeper/initial'
      })
      .get('/repos/finnp/test/issues/1234')
      .reply(200, {
        state: 'open',
        locked: false
      })
      .post('/repos/finnp/test/issues/1234/comments')
      .reply(201, () => {
        // comment added
        expect(true).toBeTruthy()
        return {}
      })

    const { repositories } = await dbs()
    const branchDoc = await repositories.get('42:branch:deadbeef')
    const repository = await repositories.get('42')

    await createInitialPrComment({
      accountId: '123',
      branchDoc,
      prDocId: '42:pr:1234',
      repository,
      combined: {state: 'success'},
      installationId: '123'
    })

    const prDoc = await repositories.get('42:pr:1234')
    expect(prDoc.initialPrCommentSent).toBeTruthy()
    await waitFor(timeToWaitAfterTests)
  })

  test('does nothing if the repo has already received the comment', async () => {
    expect.assertions(1)

    const { repositories } = await dbs()
    await upsert(repositories, '42:pr:1234', {initialPrCommentSent: true})

    nock('https://api.github.com') // eslint-disable-line
      .post('/installations/123/access_tokens')
      .optionally()
      .reply(200, {
        token: 'secret'
      })
      .get('/rate_limit')
      .optionally()
      .reply(200, {})
      .get('/repos/finnp/test')
      .reply(200, {
        name: 'test',
        html_url: 'https://github.com/finnp/test',
        newBranch: 'greenkeeper/initial'
      })
      .get('/repos/finnp/test/issues/1234')
      .reply(200, () => {
        // Should not query issue status
        expect(false).toBeFalsy()
        return {}
      })
      .post('/repos/finnp/test/issues/1234/comments')
      .reply(201, () => {
        // Should not post comment
        expect(false).toBeFalsy()
        return {}
      })

    const branchDoc = await repositories.get('42:branch:deadbeef')
    const repository = await repositories.get('42')

    const newJob = await createInitialPrComment({
      accountId: '123',
      branchDoc,
      prDocId: '42:pr:1234',
      repository,
      combined: {state: 'success'},
      installationId: '123'
    })

    expect(newJob).toBeFalsy()
    await waitFor(timeToWaitAfterTests)
  })

  test('does nothing if the issue was closed in the meanwhile', async () => {
    expect.assertions(1)

    nock('https://api.github.com') // eslint-disable-line
      .post('/installations/123/access_tokens')
      .optionally()
      .reply(200, {
        token: 'secret'
      })
      .get('/rate_limit')
      .optionally()
      .reply(200, {})
      .get('/repos/finnp/test')
      .reply(200, {
        name: 'test',
        html_url: 'https://github.com/finnp/test',
        newBranch: 'greenkeeper/initial'
      })
      .get('/repos/finnp/test/issues/1234')
      .reply(200, {
        state: 'closed',
        locked: false
      })
      .post('/repos/finnp/test/issues/1234/comments')
      .reply(201, () => {
        // Should not post comment
        expect(false).toBeFalsy()
        return {}
      })

    const { repositories } = await dbs()
    const branchDoc = await repositories.get('42:branch:deadbeef')
    const repository = await repositories.get('42')

    const newJob = await createInitialPrComment({
      accountId: '123',
      branchDoc,
      prDocId: '42:pr:1234',
      repository,
      combined: {state: 'success'},
      installationId: '123'
    })

    expect(newJob).toBeFalsy()
    await waitFor(timeToWaitAfterTests)
  })

  test('does nothing if the issue was locked in the meanwhile', async () => {
    expect.assertions(1)

    nock('https://api.github.com') // eslint-disable-line
      .post('/installations/123/access_tokens')
      .optionally()
      .reply(200, {
        token: 'secret'
      })
      .get('/rate_limit')
      .optionally()
      .reply(200, {})
      .get('/repos/finnp/test')
      .reply(200, {
        name: 'test',
        html_url: 'https://github.com/finnp/test',
        newBranch: 'greenkeeper/initial'
      })
      .get('/repos/finnp/test/issues/1234')
      .reply(200, {
        state: 'open',
        locked: true
      })
      .post('/repos/finnp/test/issues/1234/comments')
      .reply(201, () => {
        // Should not post comment
        expect(false).toBeFalsy()
        return {}
      })

    const { repositories } = await dbs()
    const branchDoc = await repositories.get('42:branch:deadbeef')
    const repository = await repositories.get('42')

    const newJob = await createInitialPrComment({
      accountId: '123',
      branchDoc,
      prDocId: '42:pr:1234',
      repository,
      combined: {state: 'success'},
      installationId: '123'
    })

    expect(newJob).toBeFalsy()
    await waitFor(timeToWaitAfterTests)
  })
})
