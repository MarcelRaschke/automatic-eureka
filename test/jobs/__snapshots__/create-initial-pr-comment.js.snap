// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-initial-pr-comment create comment for initial pr created by user 1`] = `
"
Let’s get started with automated dependency management for test :muscle:



This pull request **updates all your dependencies to their latest version**. Having them all up to date really is the best starting point for keeping up with new releases. Greenkeeper will look out for further dependency updates and make sure to handle them in isolation and in real-time, but only after **you merge this pull request**.

**Important: Greenkeeper will only start watching this repository’s dependency updates after you merge this initial pull request**.





---
<details>
<summary>🏷 How to check the status of this repository</summary>

Greenkeeper adds a badge to your README which indicates the status of this repository.

This is what your badge looks like right now :point_right:  ![Greenkeeper badge](url)
</details>
<details>
<summary>🏗 How to configure Travis CI</summary>

Greenkeeper has added a rule to your \`.travis.yml\` that whitelists Greenkeeper branches, which are created when your dependencies are updated. Travis CI will run your tests on these branches automatically to see if they still pass.

No additional setup is required 😊

</details>
<details>
<summary>🙈 How to ignore certain dependencies</summary>

You may have good reasons for not wanting to update to a certain dependency right now. In this case, you can [change the dependency’s version string in the \`package.json\` file back to whatever you prefer](https://github.com/finnp/test/edit/greenkeeper/initial/package.json).

To make sure Greenkeeper doesn’t nag you again on the next update, add a \`greenkeeper.ignore\` field to your \`package.json\`, containing a list of dependencies you don’t want to update.

\`\`\`js
// package.json
{
  …
  \\"greenkeeper\\": {
    \\"ignore\\": [
      \\"package-names\\",
      \\"you-want-me-to-ignore\\"
    ]
  }
}
\`\`\`
</details>
<details>
<summary>👩‍💻 How to update this pull request</summary>

\`\`\`bash
  # Change into your repository’s directory
  git fetch --all
  git checkout greenkeeper/initial
  npm install-test
  # Adapt your code until everything works again
  git commit -m 'chore: adapt code to updated dependencies'
  git push  greenkeeper/initial
\`\`\`
</details>
<details>
<summary>✨ How do dependency updates work with Greenkeeper?</summary>

After you merge this pull request, **Greenkeeper will create a new branch whenever a  dependency is updated**, with the new version applied. The branch creation should trigger your testing services and check whether your code still works with the new dependency version. Depending on the the results of these tests Greenkeeper will try to open meaningful and helpful pull requests and issues, so your dependencies remain working and up-to-date.

\`\`\`diff
-  \\"underscore\\": \\"^1.6.0\\"
+  \\"underscore\\": \\"^1.7.0\\"
\`\`\`

The above example shows an in-range update. \`1.7.0\` is included in the old \`^1.6.0\` range, because of the [caret \`^\` character ](https://docs.npmjs.com/misc/semver#ranges).
When the test services report success Greenkeeper will silently delete the branch again, because no action needs to be taken – everything is fine.

However, should the tests fail, Greenkeeper will create an issue to inform you about the problem immediately.

This way, you’ll never be surprised by a dependency breaking your code. As long as everything still works, Greenkeeper will stay out of your way, and as soon as something goes wrong, you’ll be the first to know.

\`\`\`diff
-  \\"lodash\\": \\"^3.0.0\\"
+  \\"lodash\\": \\"^4.0.0\\"
\`\`\`

In this example, the new version \`4.0.0\` is _not_ included in the old \`^3.0.0\` range.
For version updates like these – let’s call them “out of range” updates – you’ll receive a pull request.

This means that **you no longer need to check for new versions manually** – Greenkeeper will keep you up to date automatically.

These pull requests not only serve as reminders to update: If you have solid tests and good coverage, and the pull requests passes those tests, you can very likely just merge it and release a new version of your software straight away :shipit:

To get a better idea of which ranges apply to which releases, check out the extremely useful [semver calculator](https://semver.npmjs.com/) provided by npm.
</details>
<details>
<summary>FAQ and help</summary>

There is a collection of [frequently asked questions](https://greenkeeper.io/faq.html). If those don’t help, you can always [ask the humans behind Greenkeeper](https://github.com/greenkeeperio/greenkeeper/issues/new).
</details>


---

Good luck with your project and see you soon :sparkles:

Your [Greenkeeper](https://greenkeeper.io) bot :palm_tree:
"
`;
